design/010-sync-to-custom-resource.md

```diff
@@ -0,0 +1,113 @@
+```yaml
+---
+title: Sync to Custom Resources
+version: v1alpha1
+authors: Gustavo Carvalho
+creation-date: 2024-05-03
+status: draft
+---
+```
+
+# Sync to Custom Resources Design
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+## Summary
+
+This design document describes how `ExternalSecrets` can leverage templates to generate non-kubernetes `Secret` resource as the target. This allows to push Sensitive information to specific CRs, `ConfiMaps`, etc.
+
+## Motivation
+
+Currently, several "semi-sensitive" information needs to be provisioned directly into CRs and ConfigMaps (such as OIDC Client IDs). While these information are not strictly speaking a secret, several regulated environments must treat them as such, causing several operational overhead to deal with this information - specially on a GitOps setup.
+
+## Proposal
+
+To simplify the workflow and enhance user experience, the proposal is to integrate a functionality to template the whole manifest. this would be additional logic to the existing `target: Manifest` directly into the `templateFrom.[].target` resource. This will allow users to specify a template to render any type of manifest, instead of the original Secret object.
+
+Problems with this proposal is that the whole reconciliation logic reads from a secret - this would need to be updated if templates using Manifests are set. In that case, we should query for that specific resource as specified on the target.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+spec:
+  target:
+    name: target-custom-resource
+    manifest: # Information as to find it afterwards during reconcile cycle
+      APIVersion: my.custom.api/v1alpha1
+      Kind: CustomResource
+    template:
+      templateFrom:
+      - target: Spec #Additional target at root level (instead of data, metadata and annotations level).
```

Comment:
Not sure if we should start a naming discussion here or in the first PR, this is probably the hard part :smile: 
If the target is the root level of a YAML tree, then why not call it `target: Root` or `target: Manifest` ?
`Spec` implies that it is at the `.spec` level. If we follow the the pattern as in `Data`, `Labels` or `Annotations` this is misleading.

Reply:
I originally thought about only having this type of `target` and dump `target.manifest` bits. (and then, template everything).

The problem that I have with this is that it would be hard (if not impossible) to figure out which CR to fetch, + what is the resource name (if it is templated, a change in the secret also means we would need to do cleanup with no information on the name of the resource (complexity increases exponentially😵 💫 )). So to avoid users tampering with `apiVersion`, `Kind`, and `metadata.name`, my thought was to limit it to `Spec` - jsut because AFAIK `spec`, `metadata.annotations`, `metadata.labels` and `data` are the most common cases to template a manifest from. TBH, I'm not even sure there are others.

Having said that, for me `gjson`-like syntax might be better specially with `creationPolicy=merge` where we template only e.g. `spec.aCertOnThisPath.privateKey`